[Nguyen Hong Anh] IP

honganhcs opened the pull request:
# AnnBot
> “Life is too complicated not to be orderly.” – Martha Stewart 

AnnBot frees your mind of having to remember things you need to do. It is:
- text-based
- easy-to-learn
- ~~FAST~~ SUPER FAST to use 😉 
All you need to do is:
1. download it from [here](https://github.com/honganhcs/ip/releases)
2. double-click it
3. add your tasks 
4. let it manage your tasks for you 😄 
And it is **FREE**!

Features:
- [x] Managing tasks, deadlines and events
- [x] Searching for tasks by keywords
- [ ] Reminders (coming soon)

Here's the ` main ` method:
``` java
public static void main(String[] args) {
        new Ann().run(args);
}
```

Commit:
Add Gradle support

honganhcs committed:
Added customised greeting and echo function to Duke.

honganhcs committed:
Added the ability to store text entered by the user and display them back to the user when requested.

honganhcs committed:
Added the ability to mark tasks as done.

honganhcs committed:
Added support for tracking three types of tasks: ToDos, Deadlines and Events.

honganhcs committed:
Minor changes to Duke.java to keep the format (capitalisation) of the input.
Updated the test case.

honganhcs committed:
1. Added the ability to handle errors, including invalid commands and insufficient information.
2. Updated the test case to test the ability to handle errors.

honganhcs committed:
1. Added the ability to delete tasks from the list.
2. Updated the test case to test for the delete function.

honganhcs committed:
Added the ability to save tasks in the hard disk automatically whenever the task list changes.

honganhcs committed:
1. Updated the representation of date and time for deadlines and events.
2. Updated the test files to reflect this change.
Assumption: the user always passes in the date and time in the right format.
Improvement in the future: handle inputs in the wrong format.

honganhcs committed:
Merge branch 'branch-Level-7'

honganhcs committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

Resolution: changed the format of writing task to hard disk

honganhcs committed:
Refactored the code to extract out closely related code as classes.

honganhcs committed:
Organized the classes into suitable Java packages.

honganhcs committed:
Added JUnit tests for 2 classes: Parser and Storage.
To-do: add more JUnit tests for all public methods of all classes, including the UI component.

honganhcs committed:
Minor changes to variable name in Ann.java.

honganhcs committed:
Added the ability to find tasks that contain certain keywords.
Removed a Manifest file that was accidentally created in the previous commit.

honganhcs committed:
Add Javadoc comments for all classes.

honganhcs committed:
Change some variable names to follow Java coding standards.

honganhcs committed:
Merge branch 'branch-Level-9'

honganhcs committed:
Merge branch 'branch-A-CodingStandard'

honganhcs committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/ann/commands/AddCommand.java

Jiaaa-yang commented on src/main/java/ann/commands/AddCommand.java:
```diff
@@ -0,0 +1,60 @@
+     * @param taskComponents a String array containing information about the task.
+     */
+    private void setTask(String[] taskComponents) {
+        switch (taskType) {
+            case TODO:
+                task = new Task(taskComponents[0]);
+                break;
+            case DEADLINE:
+                task = new Deadline(taskComponents[0], taskComponents[1]);
+                break;
+            case EVENT:
+                task = new Event(taskComponents[0], taskComponents[1]);
+        }
```
The [module coding standards](https://se-education.org/guides/conventions/java/intermediate.html#layout) suggests to not indent `case` statements!

honganhcs replied:
Hi! Thanks a lot for your comments, you pointed out some issues that I overlooked and I will definitely fix them 😄

Jiaaa-yang commented on src/main/java/ann/parser/Parser.java:
```diff
@@ -0,0 +1,201 @@
+     * @return a Command.
+     */
+    public static Command parse(String input) {
+        if(input.toLowerCase().equals("bye")) {
+            return new ExitCommand();
+        } else if (input.toLowerCase().equals("list")){
+            return new ListCommand();
+        } else if (input.length() < 3) {
+            return new IncorrectCommand("Oops! Please enter a valid command!");
+        } else if(input.substring(0, 3).toLowerCase().equals("add")) {
+            return handleAdd(input.substring(3));
+        } else if(input.length() < 4) {
+            return new IncorrectCommand("Oops! Please enter a valid command!");
+        } else if(input.substring(0, 4).toLowerCase().equals("find")) {
+            return handleFind(input.substring(4));
+        } else if(input.substring(0, 4).toLowerCase().equals("mark")) {
+            return handleMark(input.substring(4));
+        } else if(input.length() < 6) {
+            return new IncorrectCommand("Oops! Please enter a valid command!");
+        } else if(input.substring(0, 6).toLowerCase().equals("unmark")) {
+            return handleUnmark(input.substring(6));
+        } else if(input.substring(0, 6).toLowerCase().equals("delete")) {
+            return handleDelete(input.substring(6));
+        } else {
+            return new IncorrectCommand("Oops! Please enter a valid command!");
+        }
```
```java
if (input.length() < 3) {
    // Check commands where 0 <= length < 3
} else if (input.length() < 4) {
    // Check commands where 3 <= length < 4
}
.
.
.
```

Do you think this portion will be more readable if it was changed to the above format? Personally I believe this block of code is not very straightforward as it seems some clauses depend on a certain length of input. Using the length as if-else conditions may help reinforce this idea.

honganhcs replied:
Actually for this, it's not just the length I have to check, but also whether it matches the command pattern, e.g. it can be shorter than 4 characters but if it is "bye" then I have to accept the command. I have to check the length in between as substring() would throw an index out of bounds exception otherwise. I'm still thinking about how this could be improved 😄

Jiaaa-yang commented on src/main/java/ann/parser/Parser.java:
```diff
@@ -0,0 +1,201 @@
+     * @return an appropriate Command object.
+     */
+    private static Command handleAdd(String input) {
+        if(input.isBlank()) {
+            return new IncorrectCommand("Oops! Please specify the task you wish to add!");
+        } else if(input.charAt(0) != ' ') {
```
The coding standard suggests a space after an `if` or `else if`!  It seems that this is followed in some places but not elsewhere, so maybe you could have overlooked it.

Jiaaa-yang commented on src/main/java/ann/storage/Storage.java:
```diff
@@ -0,0 +1,156 @@
+        return fileName.endsWith(".txt");
+    }
+
+    /**
+     * Represents an Exception thrown when a given filename is not valid.
+     */
+    public static class InvalidStorageFilePathException extends IllegalArgumentException {
+        public InvalidStorageFilePathException(String message) {
+            super(message);
+        }
+    }
```
I like that this particular exception extends from `IllegalArgumentException`! Since it also represents some form of invalid argument, and this exception also allows clearer error handling with regards to the Storage class for your AnnBot!

Jiaaa-yang commented on src/main/java/ann/commands/FindCommand.java:
```diff
@@ -0,0 +1,14 @@
+package ann.commands;
+
+public class FindCommand extends Command{
+    private String findKeyWords;
```
```suggestion
    private String keywords;
```

Do you think it would be more appropriate if this variable was renamed as above? Since 'find' is a verb which can suggest a method style of naming. This could be confusing such as in the `executeCommand` below where you pass in `this.findKeyWords` as an argument, which may seem like a method call.

honganhcs replied:
Thanks for pointing this out! I'll fix it 😄

Jiaaa-yang reviewed (commented):
Hey I think this PR is looking good! Your implementations are neat with logic properly separated into relevant classes and packages. I also like how some classes are short and do just what their name implies they should! Except for a few minor style errors, and some `if` statements with a lot of `else-if`, I think your code is highly readable!

honganhcs committed:
Change some files to follow the coding standard.

honganhcs committed:
Merge branch 'A-Gradle' into branch-A-Gradle

honganhcs committed:
Change the build.gradle file to match project.

honganhcs committed:
Fix coding style violations after running checkstyle.

honganhcs committed:
Merge branch 'branch-A-CheckStyle'

honganhcs committed:
Add a basic GUI to the app.
Things to do: 1. handle exit command, 2. implement a proper UI component

honganhcs committed:
Merge branch 'branch-Level-10'

honganhcs committed:
Change the build.gradle file to update mainClassName to Launcher.

honganhcs committed:
.gitignore, Storage class: fix bug to track files in the data folder

.gitignore includes the path data/ which is supposed to refer to the local storage folder only, but is also the name of the package containing data classes such as Task.

The common name between the local storage folder and data package causes the data package to be ignored by git as the data/ path is included in .gitignore.

Changing the name of the local storage folder and replacing the data/ path in .gitignore with the new path to the local storage folder allows git to ignore the local storage folder only, while the data package is included in the git repository.

Let's rename the local storage folder to storageFolder, replace data/ in .gitignore with storageFolder/ and assign "storageFolder" to the constant DEFAULT_STORAGE_FOLDER_NAME in the Storage class.

honganhcs committed:
Add assert statements to multiple project files

Possible bugs in the current code may lead to unexpected errors if not caught at the root.

Adding assert statements to various places in the code (e.g. in switch statements without a default case) allows the runtime to detect an assertion failure and terminate the execution with an error message.

Let's add assertion statements to places in the code which work only when certain assumptions about the program state are met.

Assertions are used instead of exceptions to check the internal logic of the code and ensure that it is correct, whereas exceptions are thrown for situations outside of the programmer's control.

honganhcs committed:
Fix bug: recover data package

The data package is missing from the repo (probably due to a bug during branching).

The data package contains essential classes such as Task which other classes depend on.

Let's copy the data package from another branch (branch-A-Assertions) into current branch.

branch-A-Assertions contains the most updated version of the data package.

honganhcs committed:
Fix code quality issues

Some method header comments are not in the recommended format, which affects the code quality.

Let's change these header comments to follow the recommended format.

The comments follow the format specified by the CS2103T website.

honganhcs committed:
Merge pull request from honganhcs/branch-A-Assertions

PR from branch-A-Assertions to master branch

honganhcs committed:
Merge branch 'master' of https://github.com/honganhcs/ip

honganhcs committed:
Merge branch 'master' into branch-A-CodeQuality

honganhcs committed:
Merge pull request from honganhcs/branch-A-CodeQuality

PR from branch-A-CodeQuality to master branch

honganhcs committed:
Merge branch 'master' of https://github.com/honganhcs/ip

honganhcs committed:
Add gradle.yml to .github/workflows

honganhcs committed:
Merge pull request from honganhcs/branch-A-CI

Add gradle.yml to .github/workflows

honganhcs committed:
Merge branch 'master' of https://github.com/honganhcs/ip

honganhcs committed:
Fix bug that causes git to track local storage folder

honganhcs committed:
Fix bug in parsing strings representing deadlines and events in local storage file

Assertion errors are thrown from the factory methods in Deadline and Event classes indicating that the strings representing deadlines and events in the local storage file do not have the correct date and time format. There is a mistake in the assert statement - it assumes that the strings in the local storage follow the format for user input when in fact the user input format and the file format are different. After fixing this mistake, another bug is discovered - the parsing of deadlines and events from the local storage file is incorrect and inadvertently adds an extra white space between the date and time.

To prevent the assertion error, the incorrect assert statements should first be removed. Then, to ensure that the deadlines and events are parsed correctly from the storage file, the relevant methods in the Deadline and Event classes should be fixed.

Let's remove the assert statements in the factory methods of Deadline and Event. Let's also change the methods used to parse the date and time of deadlines and events in the Deadline and Event classes.

The assert statements in the factory methods of Deadline and Event are temporarily removed until a good method to check if the strings in the local storage file follow the file format is found.

honganhcs committed:
TaskList class: change the find method to enable better search (partial matching of keywords)

The logic for executing the Find command is inside the find method of the TaskList class. The current find method only returns tasks that match all the inputted keywords. This limits the function of find as users may enter multiple keywords and the number of tasks that contain all those keywords in the inputted order may be few / none.

Changing the find method in the TaskList class to return tasks that match one or more of the keywords instead of all of them increases the number of matching tasks returned.

Let's change the find method in the TaskList class to add matching tasks for each inputted keyword to the returned string.

Returning the tasks that match one or more keywords improves the search quality and efficiency (users do not have to enter multiple Find commands for each keyword).

honganhcs committed:
Merge branch 'branch-C-BetterSearch'

honganhcs committed:
Add OOP to code: add a UI manager class

The Main class is in charge of initializing the GUI when the app is launched; there is no clear separation between the logic and ui components of the app.

Adding a Ui class and transferring the code from Main's start method to this new class and making AnnBot the "Main" class in charge of the app's logic ensures that there is a separation between the logic and ui components of the app.

Let's add a Ui class which manages the GUI components of the app and convert AnnBot into the Main class by letting it extend Application and implement the start method for launching the app.

Increasing the code's OOP makes it easier to maintain the code, as changes to the implementation of the UI will not affect the logic of the app.

honganhcs committed:
Changes to the controllers of the GUI components: allow users to exit the app using the command line interface (CLI)

Currently, users cannot exit from the app using the CLI as there is no method to handle a "bye" command in the MainWindow class. The "bye" command is handled in a similar way to other commands - a DialogBox containing AnnBot's response to the command is displayed on the screen.

Adding an method to handle the "bye" command in the MainWindow class will allow users to exit the app using the CLI.

Let's change the MainWindow class to handle the "bye" command differently from other commands.

honganhcs committed:
Improve the GUI: make the DialogBox expand to display the full text

honganhcs committed:
Fix bug in handleAdd method of Parser

A one-off error results in a StringIndexOutOfBoundsException being thrown when the user enters an incomplete add command e.g. (add todo).

Changing the number arguments for the substring method called inside the handleAdd method ensures that no exception is thrown.

Future improvement: getting rid of the magic numbers inside the Parser class and making the code more OOP.

honganhcs committed:
Change Parser class to add more OOP and remove magic values

There is a lot of hard code and code repetition in the Parser class which makes it difficult to change the format of the commands in the future.

Removing the magic values (numbers and Strings) and creating constants in the Command classes and the Parser class to store these values make it easier to change these values in the future.

Let's extract the arbitrary values in the code of the Parser class and assign them to constants in the Parser class and the Command classes.

honganhcs committed:
Add attributes to authors for reused and adapted code

honganhcs committed:
Improve GUI: add welcome message, adjust DialogBox's text alignment

honganhcs committed:
Merge branch 'branch-A-BetterGui'

honganhcs committed:
Add representative screenshot of product to the docs folder

honganhcs committed:
Add a representative screenshot and brief User Guide

honganhcs committed:
Add hyperlinks to easily navigate to sections

honganhcs committed:
Set theme jekyll-theme-modernist

honganhcs committed:
Merge branch 'master' of https://github.com/honganhcs/ip

honganhcs committed:
Fix Checkstyle error in Event class

honganhcs committed:
Minor change to file to improve code quality

honganhcs committed:
Update name of release

honganhcs committed:
Update README.md

honganhcs committed:
Update README.md
